reverse_replication/launcher.go

```diff
@@ -0,0 +1,273 @@
+	})
+	if err != nil {
+		fmt.Printf("Cannot submit request create change stream request: %v\n", err)
+		return
+	}
+	if err := op.Wait(ctx); err != nil {
+		if !strings.Contains(err.Error(), "Duplicate name in schema") {
+			fmt.Printf("Could not update database ddl: %v\n", err)
+			return
+		} else {
+			fmt.Printf("changestream with name '%s' already exists...skipping changestream creation\n", changeStreamName)
+		}
+	} else {
+		fmt.Println("Created changestream ", changeStreamName)
+	}
+
+	gcsclient, _ := storage.NewClient(ctx)
+	u, _ := url.Parse(sourceShardsFilePath)
+	rc, _ := gcsclient.Bucket(u.Host).Object(u.Path[1:]).NewReader(ctx)
+	bArr, _ := ioutil.ReadAll(rc)
+	rc.Close()
+	var data []interface{}
+	json.Unmarshal(bArr, &data)
+	arr := []string{}
+	for i := 0; i < len(data); i++ {
+		arr = append(arr, data[i].(map[string]interface{})["logicalShardId"].(string))
+	}
+
+	topicName := strings.Split(pubSubDataTopicId, "/")[3]
+	client, err := pubsub.NewClient(ctx, strings.Split(pubSubDataTopicId, "/")[1])
+	if err != nil {
+		fmt.Println(err)
+	}
+	defer client.Close()
+	_, err = client.CreateTopic(ctx, topicName)
+	if err != nil {
+		if !(strings.Contains(err.Error(), "AlreadyExists")) {
+			fmt.Printf("could not create topic: %v\n", err)
+			return
+		} else {
+			fmt.Printf("topic '%s' already exists, skipping creation...\n", topicName)
+		}
+	} else {
+		fmt.Println("Created topic ", pubSubDataTopicId)
+	}
+
+	wg := &sync.WaitGroup{}
+	for i := 0; i < len(arr); i++ {
+		wg.Add(1)
+		go func(shardId string) {
+			defer wg.Done()
+			_, err := client.CreateSubscription(ctx, shardId, pubsub.SubscriptionConfig{
+				Topic:                 client.Topic(topicName),
+				AckDeadline:           600 * time.Second,
+				EnableMessageOrdering: true,
+				Filter:                fmt.Sprintf("attributes.shardId=\"%s\"", shardId),
+			})
+			if err != nil {
+				if !(strings.Contains(err.Error(), "AlreadyExists")) {
+					fmt.Printf("could not create subscription: %v\n", err)
+				} else {
+					fmt.Printf("subscription '%s' already exists, skipping creation...\n", shardId)
+				}
+				return
+			}
+			fmt.Println("Created sub: ", shardId)
+		}(arr[i])
+	}
+	wg.Wait()
+
+	c, err := dataflow.NewFlexTemplatesClient(ctx)
+	if err != nil {
+		fmt.Printf("could not create flex template client: %v\n", err)
+		return
+	}
+	defer c.Close()
+
+	launchParameters := &dataflowpb.LaunchFlexTemplateParameter{
+		JobName:  fmt.Sprintf("%s-ordering", jobNamePrefix),
+		Template: &dataflowpb.LaunchFlexTemplateParameter_ContainerSpecGcsPath{ContainerSpecGcsPath: ORDERING_TEMPLATE},
+		Parameters: map[string]string{
+			"changeStreamName":   changeStreamName,
+			"instanceId":         instanceId,
+			"databaseId":         dbName,
+			"spannerProjectId":   projectId,
+			"metadataInstance":   metadataInstance,
+			"metadataDatabase":   metadataDatabase,
+			"startTimestamp":     "",
```

Comment:
I believe this is defaulted to now in the Dataflow template, might as well skip this here

Reply:
It is but keeping it for now, as if someone wants to modify this param and rerun the script, its one lesser step for them.

Reply:
There is no input parameter that takes the startTimestamp. Either remove the startTimestamp from here or take it as input parameter.